rfc/0004-at_plus.md

```diff
@@ -0,0 +1,207 @@
+The N1QL Query API specifies the following API to be used for `AT_PLUS`:
+
+```
+{"scan_vectors": {
+    "bucket1_name": {
+      "<vbucket_id>": {
+        "value": <sequence_number>,
+        "guard": "<vbucket_uuid>"
+      },
+      "<vbucket_id>": {
+        "value": <sequence_number>,
+        "guard": "<vbucket_uuid>"
+      }
+    },
+    "bucket2_name": {
+      "<vbucket_id>": {
+        "value": <sequence_number>,
+        "guard": "<vbucket_uuid>"
+      }
+    }
+  },
+  "scan_consistency": "at_plus"
+}
+```
+
+If the `Bucket` level consistency is chosen and all 1024 (or 64 on OSX) mutation tokens are available, then the shorter (and more efficient) version must be used using arrays:
+
+```
+{"scan_vectors": {
+    "bucket1_name": [
+      [<sequence_number>, "<vbucket_uuid>"], //vbucket 1
+      [<sequence_number>, "<vbucket_uuid>"], //vbucket 2
+      ...
+      [<sequence_number>, "<vbucket_uuid>"] //vbucket 1024
+    ]
+  },
+  "scan_consistency": "at_plus"
+}
+```
+
+In the 1024 vbucket case, the array form saves around 25kb (by going from `"":{"value":,"guard": ""}` to `[,]`).
+
+The `scan_vectors` object contains one entry for each bucket that takes part in this query. 
+Inside the bucket scope all `MutationTokens` that need to be part of the index boundary are put in. 
+Explicitly note that the `guard` and `vbucket_id` are strings while the `value` is a JSON number.
+
+Once the query is assembled, the Mutation Tokens are marshaled into their JSON
+representation and passed alongside all the other params to the N1QL query
+engine.
+
+How the mutation tokens are selected in the first place, depends on the type
+of query issued by the user.
+
+In its simplest form, if the user passes in the `MutationTokens` directly as part of the `Document`, they
+are encoded and the query preparation is done. In the other two cases, it is slightly
+more complicated:
+
+- If the `Document` does not contain the token but rather only the `id`, the SDK needs to apply the same
+  hashing logic as on the KV API to identify the correct vbucket ID for each
+  document ID. Once the number of IDs is known, a central Map (see below) needs
+  to be consulted which contains all the mutation tokens for every vbucket. The
+  subset is extracted and then marshalled into the query.
+- If only the `Bucket` reference is passed in, all mutation tokens for this bucket
+  (so up to 1024) need to be selected and marshalled into the query as shown
+  above. This has the highest overhead over the network.
+
+For both of these approaches to work, the SDK needs to maintain a `Token Map` for
+each open Bucket which maps a vbucket to a MutationToken. To keep this map in sync,
+after every mutation and if tokens are enabled, the map needs to be updated with
+the latest token for the touched partitions. To rephrase this slightly: the SDK
+needs to keep the state of all the mutations passed through it in the form of
+Mutation Tokens per vbucket (partition). 
```

Comment:
I'll clarify in the spec too, but you shouldn't have more than one token per vbucket, so when a new mutation arrives with a higher seqno for that vbucket you just replace it. So there are always maximum <open buckets> * <vbuckets> stored on the system.